libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -1629,6 +1714,80 @@ string ABIFunctions::extractFromStorageValue(Type const& _type, size_t _offset,
 	});
 }
 
+string ABIFunctions::calldataAccessFunction(Type const& _type)
+{
+	solAssert(_type.isValueType() || _type.dataStoredIn(DataLocation::CallData), "");
+	string functionName = "calldata_access_" + _type.identifier();
+	return createFunction(functionName, [&]() {
+		if (_type.isDynamicallyEncoded())
+		{
+			unsigned int baseEncodedSize = _type.calldataEncodedSize();
+			solAssert(baseEncodedSize > 1, "");
+			Whiskers w(R"(
+				function <functionName>(base_ref, ptr) -> <return> {
+					let rel_offset_of_tail := calldataload(ptr)
+					if iszero(slt(rel_offset_of_tail, sub(sub(calldatasize(), base_ref), sub(<neededLength>, 1)))) { revert(0, 0) }
+					value := add(rel_offset_of_tail, base_ref)
+					<handleLength>
+				}
+			)");
+			if (_type.isDynamicallySized())
+			{
+				auto const* arrayType = dynamic_cast<ArrayType const*>(&_type);
+				solAssert(!!arrayType, "");
+				unsigned int calldataStride = arrayType->calldataStride();
+				w("handleLength", Whiskers(R"(
+					length := calldataload(value)
+					value := add(value, 0x20)
+					if gt(length, 0xffffffffffffffff) { revert(0, 0) }
+					if sgt(base_ref, sub(calldatasize(), mul(length, <calldataStride>))) { revert(0, 0) }
+				)")("calldataStride", toCompactHexWithPrefix(calldataStride)).render());
+				w("return", "value, length");
+			}
+			else
+			{
+				w("handleLength", "");
+				w("return", "value");
+			}
+			w("neededLength", toCompactHexWithPrefix(baseEncodedSize));
+			w("functionName", functionName);
+			return w.render();
+		}
+		else if (_type.isValueType())
+		{
+			string decodingFunction;
+			if (auto const* functionType = dynamic_cast<FunctionType const*>(&_type))
+				decodingFunction = abiDecodingFunctionFunctionType(*functionType, false, false);
+			else
+				decodingFunction = abiDecodingFunctionValueType(_type, false);
```

Comment:
I was wondering, whether we can still do it like this after https://github.com/ethereum/solidity/pull/5815, but it should still be fine, right? We still want to validate the input, even if we just re-encode it and pass it on, so it should be good to validate and not clean here as well...

Reply:
Yes, we should validate everything that is ever read from calldata.